File: hooks/scripts/session-end.py
#!/usr/bin/env python3
"""SessionEnd hook: cleanup and final audit summary.

Generates a session summary if significant work was done.
Cleans up temporary files created during the session.
"""

import glob
import hashlib
import os
import sys
import tempfile

sys.path.insert(0, os.path.dirname(__file__))
from _lib import read_stdin, warn, passthrough, audit_log


def cleanup():
    """Remove per-user/per-project failure logs from this session."""
    uid = str(os.getuid())
    pattern = os.path.join(tempfile.gettempdir(), f"cca-failures-{uid}-*.jsonl")
    for f in glob.glob(pattern):
        try:
            os.remove(f)
        except OSError:
            pass


def count_audit_entries() -> int:
    """Count audit log entries if audit logging was enabled."""
    log_dir = os.environ.get("CCA_HOOK_LOG_DIR")
    if not log_dir:
        return 0
    log_file = os.path.join(log_dir, "cca-hooks.jsonl")
    if not os.path.exists(log_file):
        return 0
    try:
        with open(log_file) as f:
            return sum(1 for _ in f)
    except OSError:
        return 0


def main():
    read_stdin()

    audit_log("SessionEnd", "session-end.py", "session_ended")
    cleanup()

    entries = count_audit_entries()
    if entries > 0:
        warn(
            f"Session ended. {entries} hook events logged to $CCA_HOOK_LOG_DIR/cca-hooks.jsonl.",
            event="SessionEnd",
        )
    else:
        passthrough()


if __name__ == "__main__":
    main()

File: hooks/scripts/session-end.mjs
#!/usr/bin/env node
import "./suppress-stderr.mjs";
import { existsSync, readdirSync, readFileSync, unlinkSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { auditLog, passthrough, readStdin, warn } from "./_lib.mjs";

function cleanup() {
	const uid = process.getuid ? String(process.getuid()) : "unknown";
	const prefix = `cca-failures-${uid}-`;
	try {
		for (const entry of readdirSync(tmpdir())) {
			if (entry.startsWith(prefix) && entry.endsWith(".jsonl")) {
				try {
					unlinkSync(join(tmpdir(), entry));
				} catch {
					/* ignore */
				}
			}
		}
	} catch {
		// ignore
	}
}

function countAuditEntries() {
	const logDir = process.env.CCA_HOOK_LOG_DIR;
	if (!logDir) return 0;
	const logFile = join(logDir, "cca-hooks.jsonl");
	if (!existsSync(logFile)) return 0;
	try {
		return readFileSync(logFile, "utf8").split("\n").filter(Boolean).length;
	} catch {
		return 0;
	}
}

(async () => {
	try {
		await readStdin();

		auditLog("SessionEnd", "session-end.mjs", "session_ended");
		cleanup();

		const entries = countAuditEntries();
		if (entries > 0) {
			warn(
				`Session ended. ${entries} hook events logged to $CCA_HOOK_LOG_DIR/cca-hooks.jsonl.`,
				"SessionEnd",
			);
		} else {
			passthrough();
		}
	} catch {
		passthrough();
	}
})();
